File: Intro to Data Structures/pa3/Dictionary.java
//Galen Turoci
//gturoci
//Class CMPS 12B
//February 8th, 2018
//Dictionary.java takes a collection of (key, value) pairs and stores them,
//then implements the DictionaryInterface in order to allow the user to
//manipulate those pairs
//Dictionary.java

public class Dictionary implements DictionaryInterface {
	
	//private inner Node class
	private class Node {
		String key;
		String value;
		Node next;

		Node(String k, String v) {
			key = k;
			value = v;
			next = null;
		}
	}

	//Fields for the Dictionary class
	private Node head;		//reference to first Node in List
	private Node tail;		//reference to the last Node in the List
	private int numItems;	//number of items in this Dictionary
	
	//Dictionary()
	//constructor for the Dictionary class
	public Dictionary() {
		head = null;
		numItems = 0;
	}

	//private helper function -------------------------------------------------

	//findKey()
	//returns a reference to the Node at position key in this Dictionary
	private Node findKey(String key) {
		Node N = head;
		while (N != null && !key.equals(N.key)) {
			N = N.next;
			if (N == null)
				return null;
		}
		return N;
	}

	//ADT operations ----------------------------------------------------------

	//isEmpty()
	//pre: none
	//post: returns true if this Dictionary is empty, false otherwise
	public boolean isEmpty() {
		return (numItems == 0);
	}

	//size()
	//pre: none
	//post: returns the number of elements in this Dictionary
	public int size() {
		return numItems;
	}

	//lookup()
	//pre: 1 <= key <= size()
	//post: returns value at position key in this Dictionary
	public String lookup(String key) {
		Node N = head;
		while (N != null && !key.equals(N.key))
			N = N.next;
		if (N == null)
			return null;
		else
			return N.value;
	}

	//insert()
	//inserts new value into this Dictionary at position key
	//pre: key is not already in Dictionary
	//post: !isEmpty()
	public void insert(String key, String value) throws DuplicateKeyException {
		Node N = head;
		N = findKey(key);
		if (N != null)
				throw new DuplicateKeyException("Cannot insert duplicate keys.");
		if (isEmpty())
			head = new Node(key, value);
		else {
			N = head;
			while (N.next != null)
				N = N.next;
			N.next = new Node(key, value);
			tail = N.next; //sets tail to the last created Node, which is also the last Node in the list
		}
		numItems++;
	}

	//delete()
	//deletes given key from this Dictionary
	//pre: key is in Dictionary
	//post: there is one less (key, value) pair in the Dictionary
	public void delete(String key) throws KeyNotFoundException {
		Node N = findKey(key);
		if (N == null)
			throw new KeyNotFoundException("Cannot delete non-existent key.");
		if (numItems == 1)
			head = head.next;
		else if (head.key.equals(key))
			head = head.next;
		else {
			N = head;
			while (!key.equals(N.next.key))
				N = N.next;
			N.next = N.next.next;
		}
		numItems--;
	}

	//makeEmpty()
	//pre: none
	//post: isEmpty()
	public void makeEmpty() {
		head = null;
		numItems = 0;
	}

	//toString()
	//pre: none
	//post:  prints current state to stdout
	//Overrides Object's toString() method
	public String toString() {
		StringBuffer sb = new StringBuffer();
		
		for (Node N = head; N!=null; N=N.next) {
			sb.append(N.key).append(" ");
			sb.append(N.value);
			sb.append("\n");
		}
		return new String(sb);
	}

}

File: Intro to Data Structures/lab5/Dictionary.c
//Galen Turoci
//gturoci
//Class CMPS 12M
//February 15th, 2018
//Dictionary.c takes a collection of (key, value) pairs and stores them,
//then implements the DictionaryInterface in order to allow the user to
//manipulate those pairs
//DictionaryTest.c

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<assert.h>
#include"Dictionary.h"


//private types --------------------------------------------------------------

//NodeObj
typedef struct NodeObj {
	char* key;
	char* value;
	struct NodeObj* next;
} NodeObj;

//Node
typedef NodeObj* Node;

//newNode()
//constructor of the Node type
Node newNode(char* k, char* v) {
	Node N = malloc(sizeof(NodeObj));
	assert(N!=NULL);
	N->key = k;
	N->value = v;
	N->next = NULL;
	return(N);
}

//freeNode()
//destructor for the Node type
void freeNode(Node* pN) {
	if (pN!=NULL && *pN!=NULL) {
		free(*pN);
		*pN = NULL;
   }
}

//DictionaryObj
typedef struct DictionaryObj {
	Node top;
	int numItems;
} DictionaryObj;


//public functions -----------------------------------------------------------

//newDictionary()
//constructor for the Dictionary type
Dictionary newDictionary(void) {
	Dictionary D = malloc(sizeof(DictionaryObj));
	assert(D!=NULL);
	D->top = NULL;
	D->numItems = 0;
	return D;
}

//freeDictionary()
//destructor for the Dictionary type
void freeDictionary(Dictionary* pD) {
	if (pD!=NULL && *pD!=NULL) {
		if (!isEmpty(*pD))
			makeEmpty(*pD);
		free(*pD);
		*pD = NULL;
	}
}

//size()
//returns the amount of key, value pairs in this Dictionary
//pre: !isEmpty(D)
int size(Dictionary D) {
	if (D==NULL) {
		fprintf (stderr, "Dictionary Error: calling size() on NULL Dictionary reference\n");
		exit(EXIT_FAILURE);
	}
	return D->numItems;
}

//isEmpty()
//returns 1 (true) if D is empty, 0 (false) otherwise; equivalent to isEmpty
//pre: none
int isEmpty(Dictionary D) {
	if (D==NULL) {
		fprintf (stderr, "Dictionary Error: calling isEmpty() on NULL Dictionary reference\n");
		exit(EXIT_FAILURE);
	}
	return(D->numItems==0);
}

//insert()
//inserts key and value onto top of D
//pre: none
void insert(Dictionary D, char* key, char* value) {
	if (D == NULL) {
		fprintf(stderr, "Dictionary Error: calling insert() on NULL Dictionary reference\n");
		exit(EXIT_FAILURE);
	}
	if (lookup(D, key) != NULL) {
		fprintf(stderr, "Dictionary Error: Cannot insert duplicate keys\n");
		exit(EXIT_FAILURE);
	}
	if (D->top == NULL) {
		D->top = newNode(key, value);
		return;
	}
	Node N = D->top;
	while (N->next != NULL)
		N = N->next;
	N->next = newNode(key, value);
	D->numItems++;
}

//delete()
//deletes (and returns?) given key from this D
//pre: !isEmpty(D)
void delete(Dictionary D, char* key) {
	if (D==NULL) {
		fprintf(stderr, "Dictionary Error: calling delete() on NULL Dictionary reference\n");
		exit(EXIT_FAILURE);
	}
	if (D->numItems==0) {
		fprintf(stderr, "Dictionary Error: calling delete() on empty Dictionary\n");
		exit(EXIT_FAILURE);
	}
	Node N = D->top;
	if (strcmp(D->top->key, key) == 0) {
		D->top = D->top->next;
		return;
	} 
	while (strcmp(N->next->key, key) != 0) {
		N = N->next;
		if (N==NULL) {
			fprintf(stderr, "Dictionary Error: calling delete() on non-existent key\n");
			exit(EXIT_FAILURE);
		}
	}
	Node X = N->next;
	N->next = N->next->next;
	D->numItems--;
	freeNode(&X);
}

//lookup()
//returns value at position key in D //SHOULD IT RETURN KEY AND VALUE?
//pre: !isEmpty(D)
char* lookup(Dictionary D, char* key) {
	if (D == NULL) {
		fprintf(stderr, "Dictionary Error: calling lookup() on NULL Dictionary reference\n");
		exit(EXIT_FAILURE);
	}
	Node N = D->top;
	if (N == NULL)
		return NULL;
	while (strcmp(N->key, key) != 0) {
		N = N->next;
		if (N==NULL)
			return NULL;
	}
	return N->value;
}

//makeEmpty()
//resets D to the empty state
//pre: !isEmpty(D)
void makeEmpty(Dictionary D) {
	if (D==NULL) {
		fprintf(stderr, "Dictionary Error: calling makeEmpty() on NULL Dictionary reference\n");
		exit(EXIT_FAILURE);
	}
	if (D->numItems==0) {
		fprintf(stderr, "Dictionary Error: calling makeEmpty() on empty Dictionary\n");
		exit(EXIT_FAILURE);
	}
	while (D->numItems>0) {
		Node X = D->top;
		D->top = D->top->next;
		freeNode(&X);
		D->numItems--;
	}
}

//printDictionary()
//prints a text representation of D to the file pointed to by out; equivalent to toString
//pre: none
void printDictionary(FILE* out, Dictionary D) {
	Node N;
	if (D==NULL) {
		fprintf(stderr, "Dictionary Error: calling printDictionary() on NULL Dictionary reference\n");
		exit(EXIT_FAILURE);
	}
	for (N = D->top; N != NULL; N = N->next)
		fprintf(out, "%s=%s\n", N->key, N->value);
	fprintf(out, "\n");
}
